happi/audit.py

```diff
@@ -0,0 +1,423 @@
+    NO_CODE = 9
+
+
+class Command(object, metaclass=ABCMeta):
+    """
+    Bluprint for the command class
+    This will be useful if all the commands are inheriting from it
+    """
+
+    def __init__(self, name, summary):
+        self.name = name
+        self.summary = summary
+
+    @abstractmethod
+    def add_args(self, parser):
+        """
+        Adds arguments to a parser
+
+        Parameters
+        -----------
+        parser: ArgumentParser
+            Parser to add arguments to
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def run(self, args):
+        """
+        Parses the arguments given to the command.
+        And hanles the logic for this command
+
+        Parameters
+        -----------
+            args: Namespace
+        """
+        raise NotImplementedError
+
+
+class Audit(Command):
+    """
+    Audits the database for valid entries.
+    """
+    def __init__(self):
+        self.name = 'audit'
+        self.help = "Inspect a database's entries"
+        self._all_devices = set()
+        self._all_items = []
+        # 9 - NO_CODE
+        self.report_code = ReportCode(9)
+
+    def add_args(self, cmd_parser):
+        cmd_parser.add_argument(
+            "--file", help='Path to the json file (database) to be audited'
+        )
+        cmd_parser.add_argument(
+            "--extras", action='store_true', default=False,
+            help='Check specifically for extra attributes'
+        )
+
+    def run(self, args):
+        """
+        Validate the database passed in with --file option
+        """
+        if args.file is not None:
+            if self.validate_file(args.file):
+                logger.info('Using database file at %s ', args.file)
+                if args.extras:
+                    self.check_extra_attributes(args.file)
+                else:
+                    self.parse_database(args.file)
+            else:
+                logger.error('Probably provided a wrong path or filename')
+                return
+        else:
+            """
+            Validate the database defined in happi.cfg file
+            """
+            path_to_config = Client.find_config()
+
+            if path_to_config:
+                logger.info('Using Client cfg path %s ', path_to_config)
+                config = ConfigParser()
+                config.read(path_to_config)
+                try:
+                    database_path = config.get('DEFAULT', 'path')
+                except Exception as e:
+                    logger.error('Error when trying '
+                                 'to get database path %s', e)
+                    return
+                if self.validate_file(database_path):
+                    logger.info('Using database file at %s ', database_path)
+                    if args.extras:
+                        self.check_extra_attributes(database_path)
+                    else:
+                        self.parse_database(database_path)
```

Comment:
There's just a little bit of repetition here with the lines above that could be consolidated into a function or method

Reply:
Ohh thanks for pointing this out, fixed it in b183eeb.